NMS-12154: Provide a Schedule Editor to more easily create scheduled reports

mvrueden opened the pull request:
Here we add a new angular directive to more easily generate scheduled reports without having to type in a cron expression.

Jira: https://issues.opennms.org/browse/NMS-12154

# Known issues

 - [x] This PR also includes parsing a cron expression to populate the `ScheduledOptions` UI object, to only persist the cron expression and detect the UI state from that. However some corner cases are not considered properly and will result in a wrong UI state. For example only 0, 5, 10, 15, 30, 60, 120 and 180 are valid intervals, but we accept all, which means an `0/2` minute expression will result in a daily UI state with an invalid interval :-/
 - [x] Still has some TODOs

## This is how it looks in general
![Screenshot 2019-07-25 at 15 09 37](https://user-images.githubusercontent.com/4202259/61882698-4a691080-aef9-11e9-9a3f-0c80c24acfdb.png)

# Daily at 00:00
![Screenshot 2019-07-25 at 15 55 44](https://user-images.githubusercontent.com/4202259/61882716-505ef180-aef9-11e9-90cf-32552bc15152.png)

# Daily every 3 hours in the time range from 1 to 8
![Screenshot 2019-07-25 at 15 55 54](https://user-images.githubusercontent.com/4202259/61882717-50f78800-aef9-11e9-96a3-7cace684135a.png)

# Every Wednesday, Thursday and Saturday at 3:25
![Screenshot 2019-07-25 at 15 56 24](https://user-images.githubusercontent.com/4202259/61882718-50f78800-aef9-11e9-9bdb-80796609f3af.png)

# Every 1st Tuesday of the month
![Screenshot 2019-07-25 at 15 56 43](https://user-images.githubusercontent.com/4202259/61882719-50f78800-aef9-11e9-80f8-0ad766939de0.png)


## Define a cron expression manually
![Screenshot 2019-07-25 at 15 19 52](https://user-images.githubusercontent.com/4202259/61882715-505ef180-aef9-11e9-8d47-1c22229d2b10.png)


For now I left the UI in "debug mode" so you see the cron expression which is generated when you make your changes. It will also "print" the populated UI model.
Before merging, I will disable this.
![Screenshot 2019-07-25 at 16 22 04](https://user-images.githubusercontent.com/4202259/61882658-37564080-aef9-11e9-8c0c-05fa9734dc5c.png)

RangerRick commented on core/web-assets/src/main/assets/js/lib/onms-schedule-editor/templates/schedule-editor.tpl.html:
```diff
@@ -0,0 +1,169 @@
+                <select class="form-control custom-select" ng-model="options.weekOfMonth">
+                    <option value="1">first</option>
+                    <option value="2">second</option>
+                    <option value="3">third</option>
```
This year August has 5 weeks, should there be a `fourth`? I realize it's quite an edge case, but...

mvrueden replied:
Technically you can define a `fourth` or `fifth`.
However if you define, let's say the `fifth Thursday` and there is no `fifth Thursday` in that month, the report is not delivered.
From my point of view the `Schedule Editor` should make creating a schedule very easy.
If you can configure an edge case it may not be very trivial what actually happens behind the scene.
Therefore I decided to drop the `fourth` and `fifth`, just use `Last` in that scenario.
If required I can add a fourth, but would leave the fifth. 
Besides that you can always use `Cron Expression` and define everything in more detail.

> Note that if you specify “#5” and there is not 5 of the given day-of-week in the month, then no firing will occur that month.
Source: http://www.quartz-scheduler.org/documentation/quartz-2.3.0/tutorials/crontrigger.html

mvrueden commented:
@RangerRick the `Edit Button` is a place holder, as a schedule cannot be edited at this time. The functionality is just missing (-:

mvrueden commented:
Addressed reviewer feedback. The edit functionality will be implemented at a later state. See https://issues.opennms.org/browse/NMS-12150 for more details.

indigo423 commented:
I've noticed the following problem when you click the following parameters:

<img width="691" alt="Screenshot 2019-07-29 at 14 42 23" src="https://user-images.githubusercontent.com/1095181/62049286-2bc38c00-b20f-11e9-8814-0672ae0f9fca.png">

<img width="679" alt="Screenshot 2019-07-29 at 14 42 33" src="https://user-images.githubusercontent.com/1095181/62049293-2d8d4f80-b20f-11e9-8e68-f921533339c3.png">

It is not possible to schedule the report with the following error message:

> The report could not be scheduled.
> 
> **Reason:** An error occurred parsing the cron expression. It was not possible to schedule the report
> 

The seems to be the cron expression `0 0 1-14 * * TUE, THU`. As @mvrueden figured out we have to convert it to `0 0 1-14 ? * 3,5`.

Additionally: The wording in the error message is a little bit strange. "The report could not be scheduled" and "It was not possible to schedule the report" is duplicated. The last sentence doesn't have a dot :)

mvrueden commented:
@indigo423 thanks for reporting this. I looked into the report scheduling error. It seems this is a quartz limitation. I will look into this.

About the error message: Those are generated by the backend at the moment and are not optimal, but they are better than nothing. There is an issue to rework that. See https://issues.opennms.org/browse/NMS-12151 for more details.

j-white commented:
Worked well for me functionally - this is a much easier than manually editing cron expressions.

When scheduling via 'Daily' it would be nice to have the From & To fields span the entire day by default.

mvrueden commented:
@indigo423 I addressed the issue you mentioned.
@j-white I made the expression span the full day by default
@RangerRick I added the `fourth` as an option.
